syscalls.go

```diff
@@ -185,6 +186,18 @@ func bpfProgAlter(cmd int, attr *bpfProgAlterAttr) error {
 
 func bpfMapCreate(attr *bpfMapCreateAttr) (*internal.FD, error) {
 	fd, err := internal.BPF(_MapCreate, unsafe.Pointer(attr), unsafe.Sizeof(*attr))
+	if xerrors.Is(err, os.ErrPermission) {
+		// An EPERM being returned from bpf(2) could mean two things:
+		// - the user has no permissions to call bpf()
+		// - the user's rlimit is set too low to lock memory for bpf maps
+		// Since we cannot determine the exact amount of memory required,
+		// set the user's rlimit to unlimited to be able to create the map.
+		unix.Setrlimit(unix.RLIMIT_MEMLOCK, &unix.Rlimit{Cur: unix.RLIM_INFINITY, Max: unix.RLIM_INFINITY})
```

Comment:
We actually set a somewhat sensible rlimit in our application, to prevent it from mlocking all memory into place. This isn't a safe thing to do as a library.

Reply:
I agree, map creations that exceed the rlimit should return a clearer error message (afaik this throws a generic 'insufficient memory' error) and/or the fact that rlimit needs to be taken into account by the user should be better documented on the exposed API. I'll think of a better solution. :+1:

Reply:
Yeah, having a better error would be nice, thanks for taking a look. I remember that the kernel returned something generic when it ran into the rlimit, maybe that has changed?